src/Sarif/Writers/ResultLogJsonWriter.cs

```diff
@@ -167,14 +216,36 @@ public void Dispose()
                 _jsonWriter.WriteEndObject(); // End: resultsLog
             }
 
-            _writeState = State.Disposed;
+            _writeConditions |= Conditions.Disposed;
+        }
+
+        private void EnsureInitialized()
+        {
+            if (_writeConditions == Conditions.None)
+            {
+                Initialize();
+            }
+        }
+
+        private void EnsureStateNotAlreadySet(Conditions invalidConditions)
+        {
+            Conditions observedInvalidConditions = _writeConditions & invalidConditions;
+            if (observedInvalidConditions != Conditions.None)
+            {
+                // 	InvalidState	One or more invalid states were detected during serialization: {0}	
+                throw new InvalidOperationException(string.Format(SarifResources.InvalidState, observedInvalidConditions));
```

Comment:
Ok, I'm back. I see that you have some friendly error message resource strings like `RunInfoAlreadyWritten`, but you never use them. And with the existing design, where this general-purpose method `EnsureStateNotAlreadySet` reports the error, it's hard to use the friendly special-purpose error messages. This is an interesting case where our life is harder because we have a single method with two reponsibilities: detect the error, and report the error.

I would replace this method with the following:

```
private bool StateIsAlreadySet(Conditions existingConditions)
{
    return (_writeConditions & existingConditions) != Conditions.None;
}
```

Then in each of the places where you're currently calling `EnsureStateNotAlreadySet`, I'd report the appropriate friendly message. For example, in `WriteToolInfo`:

```
    // This block could be put in a helper method EnsureNotDisposed().
    if (StateIsAlreadySet(Conditions.Disposed))
    {
        throw new ObjectDisposedException("The log writer object cannot be used because it has already been disposed.");
    }

    if (StateIsAlreadySet(Conditions.ToolInfoWritten))
    {
         throw new InvalidOperationException("You cannot write ToolInfo to the log because it has already been written.");
    }
```